### PR TITLE
feat: add custom Json and JsonLd fields

### PR DESCRIPTION
**Pull request by ununhexium**
[issue 679](https://github.com/sovity/edc-extensions/issues/679)

- [x] Add custom JSON property for the asset.
- [x] Add custom JSON LD property for the asset.
- [x] ~~Add views (maps of string to string) for the JSON string properties~~

```[tasklist]
### Checklist
- [x] The PR title is short and expressive.
- [x] I have updated the CHANGELOG.md. See [changelog_update.md](https://github.com/sovity/authority-portal/tree/main/docs/dev/changelog_updates.md) for more information.
- [x] I have updated the Deployment Migration Notes Section in the CHANGELOG.md for any configuration / external API changes.
- [x] I have performed a **self-review**
```

### CODE REVIEW

File: extensions/wrapper/wrapper-common-mappers/src/main/java/de/sovity/edc/ext/wrapper/api/common/mappers/utils/UiAssetMapper.java [hunk 1]
```diff
@@ -140,19 +143,44 @@ public UiAsset buildUiAsset(JsonObject assetJsonLd, String connectorEndpoint, St
...
 
+    private String packPrivateProperties(JsonObject privateProperties) {
+        val withoutCustomJson = Json.createObjectBuilder(privateProperties)
+                .remove(Prop.SovityDcatExt.PRIVATE_CUSTOM_JSON)
+                .build();
+        val compacted = JsonLdUtils.tryCompact(withoutCustomJson);
+        return JsonUtils.toJson(compacted);
+    }
+
+    private static String packRemainingJsonLdProperties(JsonObject remaining) {
```
Comment (kulgg) [inline extensions/wrapper/wrapper-common-mappers/src/main/java/de/sovity/edc/ext/wrapper/api/common/mappers/utils/UiAssetMapper.java:176]: Maybe this should be called `packJsonLdProperties`. The context that they are remaining is independent of this function

File: extensions/wrapper/wrapper-common-mappers/src/main/java/de/sovity/edc/ext/wrapper/api/common/mappers/utils/UiAssetMapper.java [hunk 2]
```diff
@@ -140,19 +143,44 @@ public UiAsset buildUiAsset(JsonObject assetJsonLd, String connectorEndpoint, St
...
         // Private Properties
-        var privateProperties = JsonLdUtils.tryCompact(getPrivateProperties(assetJsonLd));
-        uiAsset.setPrivateProperties(getStringProperties(privateProperties));
-        uiAsset.setPrivateJsonProperties(getJsonProperties(privateProperties));
+        val privateProperties = JsonLdUtils.object(assetJsonLd, Prop.Edc.PRIVATE_PROPERTIES);
+        if (privateProperties != null) {
+            val privateCustomJson = JsonLdUtils.string(privateProperties, Prop.SovityDcatExt.PRIVATE_CUSTOM_JSON);
+            uiAsset.setPrivateCustomJsonAsString(privateCustomJson);
+
+            val privateSerializedJsonLd = packPrivateProperties(privateProperties);
```
Comment (kulgg) [inline extensions/wrapper/wrapper-common-mappers/src/main/java/de/sovity/edc/ext/wrapper/api/common/mappers/utils/UiAssetMapper.java:161]: and then we could maybe use packJsonLdProperties here and remove the `Prop.SovityDcatExt.PRIVATE_CUSTOM_JSON` from the private properties object beforehand?

And then not need the `packPrivateProperties` function

File: extensions/wrapper/wrapper-common-mappers/src/test/java/de/sovity/edc/ext/wrapper/api/common/mappers/TestUtils.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+public class TestUtils {
+    @NotNull
+    @SneakyThrows
+    public static String loadResourceAsString(String name) {
+        return new String(Files.readAllBytes(Paths.get(TestUtils.class.getResource(name).toURI())));
+    }
```
Comment (kulgg) [inline extensions/wrapper/wrapper-common-mappers/src/test/java/de/sovity/edc/ext/wrapper/api/common/mappers/TestUtils.java:14]: nice refactoring

File: extensions/wrapper/wrapper-common-mappers/src/test/java/de/sovity/edc/ext/wrapper/api/common/mappers/utils/UiAssetMapperTest.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import org.junit.jupiter.api.Test;
+
+import static net.javacrumbs.jsonunit.assertj.JsonAssertions.assertThatJson;
+import static net.javacrumbs.jsonunit.assertj.JsonAssertions.json;
+import static net.javacrumbs.jsonunit.assertj.JsonAssertions.value;
+import static org.mockito.Mockito.mock;
+
+class UiAssetMapperTest {
+    @Test
+    void test_buildAssetJsonLdContainsCustomJson() {
```
Comment (kulgg) [inline extensions/wrapper/wrapper-common-mappers/src/test/java/de/sovity/edc/ext/wrapper/api/common/mappers/utils/UiAssetMapperTest.java:16]: Very nice that you created this test class and tested the other direction.

I think the private custom json properties are missing though?

File: extensions/wrapper/wrapper-common-mappers/src/test/resources/custom-json-as-string.txt [hunk 5]
```diff
@@ -0,0 +1 @@
+{"array":[3,1,4,1,5],"boolean":false,"null":null,"number":116,"object":{"key":"value"},"string":"value"}
```
Comment (kulgg) [inline extensions/wrapper/wrapper-common-mappers/src/test/resources/custom-json-as-string.txt:1]: i think this was just a dummy file that can be removed?

File: extensions/wrapper/wrapper-common-mappers/src/main/java/de/sovity/edc/ext/wrapper/api/common/mappers/utils/UiAssetMapper.java [hunk 6]
```diff
@@ -236,14 +264,18 @@ private JsonObjectBuilder getAssetProperties(
     private JsonObjectBuilder getAssetPrivateProperties(UiAssetCreateRequest uiAssetCreateRequest) {
         var privateProperties = Json.createObjectBuilder();
 
-        var stringProperties = uiAssetCreateRequest.getPrivateProperties();
```
Comment (kulgg) [inline extensions/wrapper/wrapper-common-mappers/src/main/java/de/sovity/edc/ext/wrapper/api/common/mappers/utils/UiAssetMapper.java:239]: the `getPrivateProperties` function can be removed as well

Comment (kulgg) [reply extensions/wrapper/wrapper-common-mappers/src/main/java/de/sovity/edc/ext/wrapper/api/common/mappers/utils/UiAssetMapper.java:239]: `getStringProperties` and `getJsonProperties` are also unused

File: extensions/wrapper/wrapper/src/test/java/de/sovity/edc/ext/wrapper/api/ui/pages/asset/AssetApiServiceTest.java [hunk 7]
```diff
@@ -132,6 +133,28 @@ void testAssetCreation(AssetService assetService) {
...
                 .publisherHomepage("publisherHomepage")
                 .dataAddressProperties(dataAddressProperties)
+                .customJsonAsString("{\"test\":\"value\"}")
+                .customJsonLdAsString("""
+                        {
+                            "https://string": "value",
+                            "https://number": 3.14,
+                            "https://array": [1,2,3],
+                            "https://object": { "https://key": "value" },
+                            "https://booleans/are/not/supported/by/Eclipse/EDC": true,
```
Comment (kulgg) [inline extensions/wrapper/wrapper/src/test/java/de/sovity/edc/ext/wrapper/api/ui/pages/asset/AssetApiServiceTest.java:143]: lol

File: extensions/wrapper/wrapper/src/test/java/de/sovity/edc/ext/wrapper/api/ui/pages/contract_agreement/ContractAgreementPageTest.java [hunk 8]
```diff
@@ -82,7 +82,7 @@ void testContractAgreementPage(
         // arrange
         assetIndex.create(asset(ASSET_ID)).orElseThrow(storeFailure -> new RuntimeException("Failed to create asset"));
         contractNegotiationStore.save(contractDefinition(CONTRACT_DEFINITION_ID));
-        transferProcessStore.updateOrCreate(transferProcess(1, 1, TransferProcessStates.COMPLETED.code()));
+        transferProcessStore.save(transferProcess(1, 1, TransferProcessStates.COMPLETED.code()));
```
Comment (kulgg) [inline extensions/wrapper/wrapper/src/test/java/de/sovity/edc/ext/wrapper/api/ui/pages/contract_agreement/ContractAgreementPageTest.java:85]: why the change?

File: tests/src/test/java/de/sovity/edc/e2e/Ms8ConnectorMigrationTest.java [hunk 9]
```diff
@@ -124,15 +123,6 @@ void testMs8DataOffer_Properties() {
             softly.assertThat(asset.getPublisherHomepage()).isEqualTo("https://publisher");
             softly.assertThat(asset.getTransportMode()).isEqualTo("Rail");
             softly.assertThat(asset.getVersion()).isEqualTo("1.0");
-            softly.assertThat(asset.getAdditionalProperties()).isEqualTo(Map.of(
-                    "http://unknown/usecase", "my-use-case",
-                    "http://unknown/custom-prop-1", "1",
-                    "http://custom-prop-2", "2",
-                    "https://custom-prop-3", "3"
-            ));
```
Comment (kulgg) [inline tests/src/test/java/de/sovity/edc/e2e/Ms8ConnectorMigrationTest.java:132]: Should we add a customJsonString assertion to replace this? and keep the fields in the migration file?

File: tests/src/test/java/de/sovity/edc/e2e/UiApiWrapperTest.java [hunk 10]
```diff
@@ -403,12 +421,15 @@ void editAssetMetadataOnLiveContract() {
         // act
         providerClient.uiApi().editAssetMetadata(assetId, UiAssetEditMetadataRequest.builder()
                 .title("Good Asset Title")
+                .customJsonAsString("{\"edited\":\"new value\"}")
                 .build());
```
Comment (kulgg) [inline tests/src/test/java/de/sovity/edc/e2e/UiApiWrapperTest.java:425]: we could also add customJsonLdAsString to the test

**Review by kulgg — CHANGES_REQUESTED** [review]

**Review by ununhexium — COMMENTED** [review]
Needs improvements

File: UPDATES.md [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+We must therefore re-build the asset using the same content as that `.toBuilder()`.
+
+If the Eclipse EDC ads a field in this builder, we will miss it and any write to the JsonLd via the web API
+will remove that hypothetical new field.
+
+## Workaround
+
+On the EDC version update, check that `org.eclipse.edc.spi.types.domain.asset.Asset.toBuilder` doesn't set more
+fields than what we set. If a new field was added, add it to this function too.
```
[the remark this answers is not in the document]
Comment (kulgg) [reply UPDATES.md:20]: Okay makes sense

[the remark this answers is not in the document]
Comment (kulgg) [reply UPDATES.md:20]: Maybe we should describe at the top of UPDATE.md what this file is for and then format it more like a list (even if there is only one point right now)

**Review by kulgg — DISMISSED** [review]
LGTM :)

File: UPDATES.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (ununhexium) [reply UPDATES.md:20]: agree

File: extensions/wrapper/wrapper-common-mappers/src/main/java/de/sovity/edc/ext/wrapper/api/common/mappers/utils/UiAssetMapper.java (hunk 6 above)
Comment (ununhexium) [reply extensions/wrapper/wrapper-common-mappers/src/main/java/de/sovity/edc/ext/wrapper/api/common/mappers/utils/UiAssetMapper.java:239]: Cleaned up.

File: extensions/wrapper/wrapper/src/test/java/de/sovity/edc/ext/wrapper/api/ui/pages/asset/AssetApiServiceTest.java [hunk 12]
```diff
@@ -281,6 +346,14 @@ void testEditAssetMetadata(AssetService assetService) {
         assertThat(asset.getHttpDatasourceHintsProxyPath()).isTrue();
         assertThat(asset.getHttpDatasourceHintsProxyQueryParams()).isTrue();
         assertThat(asset.getHttpDatasourceHintsProxyBody()).isTrue();
+        assertThat(asset.getCustomJsonAsString()).isEqualTo("""
+                { "edited": "new value" }
+                """);
+        assertThatJson(asset.getCustomJsonLdAsString()).isEqualTo("""
```
[the remark this answers is not in the document]
Comment (ununhexium) [reply extensions/wrapper/wrapper/src/test/java/de/sovity/edc/ext/wrapper/api/ui/pages/asset/AssetApiServiceTest.java:352]: Yes. Done.

File: UPDATES.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (ununhexium) [reply UPDATES.md:20]: added.

File: UPDATES.md [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
```
Comment (kulgg) [inline UPDATES.md:25]: Maybe we can reformulate it like this

```
# Updates checklist

After each EDC version update

- [ ]  Check if `org.eclipse.edc.spi.types.domain.asset.Asset.toBuilder` added a new
field and adjust the builder in `de.sovity.edc.ext.wrapper.api.ui.pages.asset.AssetBuilder.fromEditMetadataRequest` accordingly

## Context

### Asset.toBuilder

A list of the element that may break when updating the EDC version.

In `de.sovity.edc.ext.wrapper.api.ui.pages.asset.AssetBuilder.fromEditMetadataRequest`

When re-creating the asset, we can't re-use the `Asset.toBuilder()` as it doesn't allow us to remove properties.

We must therefore re-build the asset using the same content as that `.toBuilder()`.

If the Eclipse EDC adds a field in this builder, we will miss it and any write to the JsonLd via the web API
will remove that hypothetical new field.

#### Workaround

On the EDC version update, check that `org.eclipse.edc.spi.types.domain.asset.Asset.toBuilder` doesn't set more
fields than what we set. If a new field was added, add it to this function too.
```

**Review by kulgg — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 14]
```diff
@@ -15,10 +15,16 @@ please see [changelog_updates.md](docs/dev/changelog_updates.md).
...
 
+The `additionalProperties`, `additionalJsonProperties`, `privateProperties` and `privateJsonProperties` have been removed.
+
+They are replaced with the more generic `customJson`, `customJsonLd`, `privateCustomJson` and `privateCustomJsonLd`.
+
 #### Patch Changes
 
 ### Deployment Migration Notes
 
+Check that you don't use the old `additionalProperties` fields anymore, see Minor Changes for details.
```
[the remark this answers is not in the document]
Comment (ununhexium) [reply CHANGELOG.md:26]: ok

File: CHANGELOG.md [hunk 15]
```diff
@@ -15,10 +15,16 @@ please see [changelog_updates.md](docs/dev/changelog_updates.md).
 
 #### Minor Changes
 
+The `additionalProperties`, `additionalJsonProperties`, `privateProperties` and `privateJsonProperties` have been removed.
+
+They are replaced with the more generic `customJson`, `customJsonLd`, `privateCustomJson` and `privateCustomJsonLd`.
+
```
[the remark this answers is not in the document]
Comment (ununhexium) [reply CHANGELOG.md]: Fixed

File: UPDATES.md (hunk 13 above)
Comment (ununhexium) [reply UPDATES.md]: merged

**Review by kulgg — APPROVED** [review]
LGTM
